npm/pkg/dataplane/dataplane.go

```diff
@@ -364,18 +393,69 @@ func (dp *DataPlane) applyDataPlaneNow(context string) error {
 func (dp *DataPlane) AddPolicy(policy *policies.NPMNetworkPolicy) error {
 	klog.Infof("[DataPlane] Add Policy called for %s", policy.PolicyKey)
 
-	// Create and add references for Selector IPSets first
-	err := dp.createIPSetsAndReferences(policy.AllPodSelectorIPSets(), policy.PolicyKey, ipsets.SelectorType)
-	if err != nil {
-		klog.Infof("[DataPlane] error while adding Selector IPSet references: %s", err.Error())
-		return fmt.Errorf("[DataPlane] error while adding Selector IPSet references: %w", err)
+	if !dp.netPolInBackground {
+		return dp.addPolicies([]*policies.NPMNetworkPolicy{policy})
 	}
 
-	// Create and add references for Rule IPSets
-	err = dp.createIPSetsAndReferences(policy.RuleIPSets, policy.PolicyKey, ipsets.NetPolType)
-	if err != nil {
-		klog.Infof("[DataPlane] error while adding Rule IPSet references: %s", err.Error())
-		return fmt.Errorf("[DataPlane] error while adding Rule IPSet references: %w", err)
+	// Choose to keep netPolQueue locked while running iptables-restore within addPoliciesWithRetry.
+	// We are not blocking any thread but the background NetPol thread, which would run the same command anyways
+	dp.netPolQueue.Lock()
+	defer dp.netPolQueue.Unlock()
+
+	dp.netPolQueue.enqueue(policy)
+	newCount := dp.netPolQueue.len()
+
+	klog.Infof("[DataPlane] [%s] new pending netpol count: %d", contextAddNetPol, newCount)
+
+	if newCount >= dp.MaxPendingNetPols {
+		klog.Infof("[DataPlane] [%s] applying now since reached maximum batch count: %d", contextAddNetPol, newCount)
+		dp.addPoliciesWithRetry(contextAddNetPol)
+	}
+	return nil
+}
+
+// addPoliciesWithRetry tries adding all policies. If this fails, it tries adding policies one by one.
+// The caller must lock netPolQueue.
+func (dp *DataPlane) addPoliciesWithRetry(context string) {
+	netPols := dp.netPolQueue.dump()
```

Comment:
storing a queue of netpols as a map in background, but when applied we convert from map back to slice, any reason to not store background queue as a slice and then avoid this conversion?

Reply:
having the map will help with `dp.RemovePolicy()`, where we remove the NetPol from the queue